packages/core/src/sync/index.ts

```diff
@@ -566,180 +579,241 @@ export const createSync = async (args: CreateSyncParameters): Promise<Sync> => {
+        syncProgress.finalized = event.block;
+        const checkpoint = getOmnichainCheckpoint("finalized")!;
+
+        // Raise event to parent function (runtime)
+        if (checkpoint > prev) {
+          args.onRealtimeEvent({ type: "finalize", checkpoint });
+        }
+
+        const finalizedEventData = unfinalizedEventData.filter(
+          (ued) =>
+            hexToNumber(ued.block.number) <= hexToNumber(event.block.number),
+        );
+
+        localSyncContext.get(network)!.unfinalizedEventData =
+          unfinalizedEventData.filter(
+            (ued) =>
+              hexToNumber(ued.block.number) > hexToNumber(event.block.number),
           );
 
-          await args.syncStore.pruneByBlock({
-            blocks: event.reorgedBlocks,
+        if (
+          getChainCheckpoint({ syncProgress, network, tag: "finalized" })! >
+          getOmnichainCheckpoint("current")!
+        ) {
+          args.common.logger.warn({
+            service: "sync",
+            msg: `Finalized block for '${network.name}' has surpassed overall indexing checkpoint`,
+          });
+        }
+
+        // Add finalized blocks, logs, transactions, receipts, and traces and intervals to the sync-store.
+
+        await Promise.all([
+          args.syncStore.insertBlocks({
+            blocks: finalizedEventData
+              .filter(({ filters }) => filters.size > 0)
+              .map(({ block }) => block),
+            chainId: network.chainId,
+          }),
+          args.syncStore.insertLogs({
+            logs: finalizedEventData.flatMap(({ logs, block }) =>
+              logs.map((log) => ({ log, block })),
+            ),
+            shouldUpdateCheckpoint: true,
+            chainId: network.chainId,
+          }),
+          args.syncStore.insertLogs({
+            logs: finalizedEventData.flatMap(({ factoryLogs }) =>
+              factoryLogs.map((log) => ({ log })),
+            ),
+            shouldUpdateCheckpoint: false,
+            chainId: network.chainId,
+          }),
+          args.syncStore.insertTransactions({
+            transactions: finalizedEventData.flatMap(
+              ({ transactions }) => transactions,
+            ),
             chainId: network.chainId,
+          }),
+          args.syncStore.insertTransactionReceipts({
+            transactionReceipts: finalizedEventData.flatMap(
+              ({ transactionReceipts }) => transactionReceipts,
+            ),
+            chainId: network.chainId,
+          }),
+          args.syncStore.insertCallTraces({
+            callTraces: finalizedEventData.flatMap(({ callTraces, block }) =>
+              callTraces.map((callTrace) => ({ callTrace, block })),
+            ),
+            chainId: network.chainId,
+          }),
+          ...args.sources
+            .filter(({ filter }) => filter.chainId === network.chainId)
+            .map(({ filter }) =>
+              args.syncStore.insertInterval({ filter, interval }),
+            ),
```

Comment:
Really nice that this is now happening in the same spot - not necessary for this PR, but ultimately I think it would be great to use a database transaction here.